src/AggregateFunctions/AggregateFunctionSequenceNextNode.h

```diff
@@ -0,0 +1,483 @@
+
+        const UInt64 new_elems = std::min(data(rhs).value.size(), static_cast<size_t>(max_elems) - data(place).value.size());
+        for (UInt64 i = 0; i < new_elems; ++i)
+            a.push_back(b[i]->clone(arena), arena);
+
+        /// Either sort whole container or do so partially merging ranges afterwards
+        using Comparator = typename SequenceNextNodeGeneralData<Node, Descending>::Comparator;
+
+        if (!data(place).sorted && !data(rhs).sorted)
+            std::stable_sort(std::begin(a), std::end(a), Comparator{});
+        else
+        {
+            const auto begin = std::begin(a);
+            const auto middle = std::next(begin, a_size);
+            const auto end = std::end(a);
+
+            if (!data(place).sorted)
+                std::stable_sort(begin, middle, Comparator{});
+
+            if (!data(rhs).sorted)
+                std::stable_sort(middle, end, Comparator{});
+
+            std::inplace_merge(begin, middle, end, Comparator{});
+        }
+
+        data(place).sorted = true;
+    }
+
+    void serialize(ConstAggregateDataPtr __restrict place, WriteBuffer & buf) const override
+    {
+        writeBinary(data(place).sorted, buf);
+
+        auto & value = data(place).value;
+        writeVarUInt(value.size(), buf);
+        for (auto & node : value)
+            node->write(buf);
+    }
+
+    void deserialize(AggregateDataPtr __restrict place, ReadBuffer & buf, Arena * arena) const override
+    {
+        readBinary(data(place).sorted, buf);
+
+        UInt64 size;
+        readVarUInt(size, buf);
+
+        if (unlikely(size == 0))
+            return;
+
+        auto & value = data(place).value;
+
+        value.resize(size, arena);
+        for (UInt64 i = 0; i < size; ++i)
+            value[i] = Node::read(buf, arena);
+    }
+
+    /// Calculate position of current event in target chain and shift to corresponding offset
+    /// Lets consider case where we search chain 'ABCD':
+    /// - If current event is 'X' we can skip it and perform next step from position after this 'X'
+    /// - If current event is 'A' we will start from this position
+    /// - If current event is 'B' then second position in our chain should match this 'B'.
+    ///   And we perform next step from position one before 'B'.
+    /// - And so on...
+    inline UInt32 calculateJump(const Data & data, const UInt32 i, const UInt32 j) const
+    {
+        /// Fast check if value is zero, not in sequence
+        if (data.value[i - j]->events_bitset.none())
+            return events_size - j;
+
+        UInt32 k = 1;
+        for (; k < events_size - j; ++k)
+            if (data.value[i - j]->events_bitset.test(events_size - 1 - j - k))
+                break;
+        return k;
+    }
+
+    /// This method returns an index of next node that matched the events.
+    /// It is one as referring Boyer-Moore-Algorithm(https://en.wikipedia.org/wiki/Boyer%E2%80%93Moore_string-search_algorithm).
```

Comment:
Important part of Boyer-Moore is shift table for bad character. But we do not build it explicitly because we know position of each event and can determinate it from flag in bitset. Is it right?

Reply:
That's right.